src/soc/amd/common/block/spi/fch_spi_ctrl.c

```diff
@@ -217,76 +247,129 @@ static int fch_spi_flash_protect(const struct spi_flash *flash, const struct reg
 				 const enum ctrlr_prot_type type)
 {
 	int ret;
-	u32 reg32, rom_base, range_base;
-	size_t addr, len, gran_value, total_ranges, range;
+	u8 reg8;
+	u32 reg32, rom_base, flags;
+	size_t addr, len, gran_value, total_ranges, range, protected_ranges;
 	bool granularity_64k = true; /* assume 64k granularity */
+	struct flash_mmap_window mmap_window;
 
 	addr = region->offset;
 	len = region->size;
 
+	/* Multiple flash windows not supported right now */
+	if (spi_flash_get_mmap_windows(&mmap_window) != 1)
+		return -1;
+
 	reg32 = pci_read_config32(SOC_LPC_DEV, ROM_ADDRESS_RANGE2_START);
 	rom_base = WORD_TO_DWORD_UPPER(reg32);
-	if (addr < rom_base)
-		return -1;
-	range_base = addr % rom_base;
+
+	/* Normalize against host memory mapped base of flash */
+	if (addr >= mmap_window.flash_base && addr < (mmap_window.flash_base + mmap_window.size)) {
+		if ((mmap_window.host_base + addr) < rom_base) {
+			printk(BIOS_ERR, "FCH SPI: ROM protect base lower than ROM base\n");
+			return -1;
+		}
+		rom_base += addr;
+	} else {
+		if (addr < rom_base) {
+			printk(BIOS_ERR, "FCH SPI: ROM protect base lower than ROM base\n");
+			return -1;
+		}
+		rom_base = addr;
+	}
 
 	/* Define granularity to be used */
-	if (GRANULARITY_TEST_4k & range_base)
+	if (GRANULARITY_TEST_4k & rom_base)
 		granularity_64k = false; /* use 4K granularity */
 	if (GRANULARITY_TEST_4k & len)
 		granularity_64k = false; /* use 4K granularity */
 
 	/* Define the first range and total number of ranges required */
-	if (granularity_64k) {
-		gran_value = 0x00010000; /* 64 KiB */
-		range_base = range_base >> 16;
-	} else {
-		gran_value = 0x00001000; /* 4 KiB */
-		range_base = range_base >> 12;
-	}
+	if (granularity_64k)
+		gran_value = 64*KiB; /* 64 KiB */
+	else
+		gran_value = 4*KiB; /* 4 KiB */
+
 	total_ranges = len / gran_value;
-	range_base &= RANGE_ADDR_MASK;
+
+	if (total_ranges == 0) {
+		printk(BIOS_ERR, "fch_spi_flash_protect: Incorrect protection range size\n");
+		return -1;
+	}
+
+	/* Range equal zero means a protection of a single block of 64KiB or 4KiB */
+	total_ranges--;
```

Comment:
This applies to every register, but it is subtracted only once. Take for example `total_ranges = 0x100` (after decrement) and try following the code.

Reply:
Right, I have to subtract it each time the next register is programmed too.